XenAdmin/SettingsPanels/NRPEEditPage.cs

```diff
@@ -0,0 +1,397 @@
+            CheckGroupList.Add(new CheckGroup("check_host_cpu", Messages.NRPE_CHECK_HOST_CPU));
+            CheckGroupList.Add(new CheckGroup("check_host_memory", Messages.NRPE_CHECK_HOST_MEMORY));
+            CheckGroupList.Add(new CheckGroup("check_vgpu", Messages.NRPE_CHECK_VGPU));
+            CheckGroupList.Add(new CheckGroup("check_vgpu_memory", Messages.NRPE_CHECK_VGPU_MEMORY));
+            CheckGroupList.Add(new Dom0LoadCheckGroup("check_load", Messages.NRPE_CHECK_LOAD));
+            CheckGroupList.Add(new CheckGroup("check_cpu", Messages.NRPE_CHECK_CPU));
+            CheckGroupList.Add(new CheckGroup("check_memory", Messages.NRPE_CHECK_MEMORY));
+            CheckGroupList.Add(new FreeCheckGroup("check_swap", Messages.NRPE_CHECK_SWAP));
+            CheckGroupList.Add(new FreeCheckGroup("check_disk_root", Messages.NRPE_CHECK_DISK_ROOT));
+            CheckGroupList.Add(new FreeCheckGroup("check_disk_log", Messages.NRPE_CHECK_DISK_LOG));
+
+            foreach (CheckGroup checkGroup in CheckGroupList)
+            {
+                CheckDataGridView.Rows.Add(checkGroup.CheckThresholdRow);
+                CheckGroupDictByName.Add(checkGroup.Name, checkGroup);
+                CheckGroupDictByLabel.Add(checkGroup.NameCell.Value.ToString(), checkGroup);
+            }
+
+            UpdateOtherComponentBasedEnableNRPECheckBox(false);
+
+            AllowHostsTextBox.ForeColor = Color.FromKnownColor(KnownColor.ControlDark);
+            AllowHostsTextBox.GotFocus += AllowHostsTextBox_GotFocus;
+            AllowHostsTextBox.LostFocus += AllowHostsTextBox_LostFocus;
+
+            if (isHost)
+            {
+                PoolTipsPicture.Hide();
+                PoolTipsLabel.Hide();
+            }
+            InvalidParamToolTip = new ToolTip
+            {
+                IsBalloon = true,
+                ToolTipIcon = ToolTipIcon.Warning,
+                ToolTipTitle = Messages.INVALID_PARAMETER,
+                Tag = AllowHostsTextBox
+            };
+        }
+
+        public bool ValidToSave
+        {
+            get
+            {
+                InvalidParamToolTipText = "";
+                InvalidParamToolTip.ToolTipTitle = "";
+                CheckDataGridView.ShowCellToolTips = false;
+
+                if (!EnableNRPECheckBox.Checked)
+                {
+                    return true;
+                }
+
+                bool valid = true;
+                if (!IsAllowHostsValid())
+                {
+                    valid = false;
+                }
+
+                foreach (CheckGroup checkGroup in CheckGroupList)
+                {
+                    if (!checkGroup.CheckValue())
+                    {
+                        valid = false;
+                    }
+                }
+                return valid;
+            }
+        }
+
+        public bool HasChanged
+        {
+            get
+            {
+                return true;

```

Comment:
It is not guaranteed that the user will change something. They may click OK without any changes in this dialog or they becasue they have changed something in a different panel. In this case we should not run the action to save changes unnecessarily